File: tablet/Resources/task/Gad7.js
// Gad7.js

/*
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

/*jslint node: true, newcap: true, nomen: true, plusplus: true */
"use strict";
/*global Titanium, L */

var DBCONSTANTS = require('common/DBCONSTANTS'),
    dbcommon = require('lib/dbcommon'),
    taskcommon = require('lib/taskcommon'),
    lang = require('lib/lang'),
    tablename = "gad7",
    fieldlist = dbcommon.standardTaskFields(),
    nquestions = 7;

dbcommon.appendRepeatedFieldDef(fieldlist, "q", 1, nquestions,
                                DBCONSTANTS.TYPE_INTEGER);

// CREATE THE TABLE

dbcommon.createTable(tablename, fieldlist);

// TASK

function Gad7(patient_id) {
    taskcommon.BaseTask.call(this, patient_id); // call base constructor
}

lang.inheritPrototype(Gad7, taskcommon.BaseTask);
lang.extendPrototype(Gad7, {

    // KEY DATABASE FIELDS (USED BY DatabaseObject)

    _objecttype: Gad7,
    _tablename: tablename,
    _fieldlist: fieldlist,

    // TASK CLASS FIELD OVERRIDES (USED BY BaseTask)

    // OTHER

    // Scoring
    getTotalScore: function () {
        return taskcommon.totalScoreByPrefix(this, "q", 1, nquestions);
    },

    // Standard task functions
    isComplete: function () {
        return taskcommon.isCompleteByPrefix(this, "q", 1, nquestions);
    },

    getSummary: function () {
        return (L('total_score') + " " + this.getTotalScore() + "/21" +
                this.isCompleteSuffix());
    },

    getDetail: function () {
        var totalscore = this.getTotalScore(),
            severity = (totalscore >= 15 ? L('severe')
                        : (totalscore >= 10 ? L('moderate')
                           : (totalscore >= 5 ? L('mild')
                              : L('none')
                             )
                          )
                       );
        // Scoring: ref http://www.phqscreeners.com/instructions/instructions.pdf
        return (
            taskcommon.valueDetail(this, "gad7_q", "_s", " ", "q", 1,
                                   nquestions) +
            "\n" +
            this.getSummary() + "\n" +
            "\n" +
            L('gad7_anxiety_severity') + " " + severity
        );
    },

    edit: function (readOnly) {
        var self = this,
            KeyValuePair = require('lib/KeyValuePair'),
            Questionnaire = require('questionnaire/Questionnaire'),
            pages,
            questionnaire;

        pages = [
            {
                title: L('gad7_title'),
                elements: [
                    { type: "QuestionText", text: L('gad7_stem') },
                    {
                        type: "QuestionMCQGrid",
                        options: [
                            new KeyValuePair(L('gad7_a0'), 0),
                            new KeyValuePair(L('gad7_a1'), 1),
                            new KeyValuePair(L('gad7_a2'), 2),
                            new KeyValuePair(L('gad7_a3'), 3)
                        ],
                        questions: [
                            L('gad7_q1'),
                            L('gad7_q2'),
                            L('gad7_q3'),
                            L('gad7_q4'),
                            L('gad7_q5'),
                            L('gad7_q6'),
                            L('gad7_q7')
                        ],
                        fields: [ 'q1', 'q2', 'q3', 'q4', 'q5', 'q6', 'q7']
                    }
                ]
            }
        ];

        questionnaire = new Questionnaire({
            readOnly: readOnly,
            pages: pages,
            callbackThis: self,
            fnGetFieldValue: self.defaultGetFieldValueFn,
            fnSetField: self.defaultSetFieldFn,
            fnFinished: self.defaultFinishedFn
        });
        questionnaire.open();
    }

});

module.exports = Gad7;

File: server/tasks/gad7.py
#!/usr/bin/env python3
# gad7.py

"""
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
"""

from ..cc_modules.cc_constants import (
    CTV_DICTLIST_INCOMPLETE,
)
from ..cc_modules.cc_db import repeat_fieldspec
from ..cc_modules.cc_html import (
    answer,
    tr,
    tr_qa,
)
from ..cc_modules.cc_string import WSTRING
from ..cc_modules.cc_task import get_from_dict, Task


# =============================================================================
# GAD-7
# =============================================================================

class Gad7(Task):
    NQUESTIONS = 7

    tablename = "gad7"
    shortname = "GAD-7"
    longname = "Generalized Anxiety Disorder Assessment"
    fieldspecs = repeat_fieldspec(
        "q", 1, NQUESTIONS, min=0, max=3,
        comment_fmt="Q{n}, {s} (0 not at all - 3 nearly every day)",
        comment_strings=[
            "nervous/anxious/on edge",
            "can't stop/control worrying",
            "worrying too much about different things",
            "trouble relaxing",
            "restless",
            "irritable",
            "afraid"
        ])

    TASK_FIELDS = [x["name"] for x in fieldspecs]

    def get_trackers(self):
        return [
            {
                "value": self.total_score(),
                "plot_label": "GAD-7 total score",
                "axis_label": "Total score (out of 21)",
                "axis_min": -0.5,
                "axis_max": 21.5,
                "horizontal_lines": [
                    14.5,
                    9.5,
                    4.5
                ],
                "horizontal_labels": [
                    (17, WSTRING("severe")),
                    (12, WSTRING("moderate")),
                    (7, WSTRING("mild")),
                    (2.25, WSTRING("none"))
                ]
            }
        ]

    def get_clinical_text(self):
        if not self.is_complete():
            return CTV_DICTLIST_INCOMPLETE
        return [{"content": "GAD-7 total score {}/21 ({})".format(
            self.total_score(), self.severity())}]

    def get_summaries(self):
        return [
            self.is_complete_summary_field(),
            dict(name="total", cctype="INT", value=self.total_score(),
                 comment="Total score (/21)"),
            dict(name="severity", cctype="TEXT", value=self.severity(),
                 comment="Severity"),
        ]

    def is_complete(self):
        return (
            self.are_all_fields_complete(self.TASK_FIELDS) and
            self.field_contents_valid()
        )

    def total_score(self):
        return self.sum_fields(self.TASK_FIELDS)

    def severity(self):
        score = self.total_score()
        if score >= 15:
            severity = WSTRING("severe")
        elif score >= 10:
            severity = WSTRING("moderate")
        elif score >= 5:
            severity = WSTRING("mild")
        else:
            severity = WSTRING("none")
        return severity

    def get_task_html(self):
        score = self.total_score()
        severity = self.severity()
        answer_dict = {None: None}
        for option in range(0, 4):
            answer_dict[option] = (
                str(option) + " — " + WSTRING("gad7_a" + str(option))
            )
        h = """
            <div class="summary">
                <table class="summary">
        """ + self.get_is_complete_tr()
        h += tr(WSTRING("total_score"), answer(score) + " / 21")
        h += tr(WSTRING("gad7_anxiety_severity") + " <sup>[1]</sup>",
                severity)
        h += """
                </table>
            </div>
            <div class="explanation">
                Ratings are over the last 2 weeks.
            </div>
            <table class="taskdetail">
                <tr>
                    <th width="50%">Question</th>
                    <th width="50%">Answer</th>
                </tr>
        """
        for q in range(1, self.NQUESTIONS + 1):
            h += tr_qa(
                WSTRING("gad7_q" + str(q)),
                get_from_dict(answer_dict, getattr(self, "q" + str(q)))
            )
        h += """
            </table>
            <div class="footnotes">
                [1] ≥15 severe, ≥10 moderate, ≥5 mild.
                Score ≥10 identifies: generalized anxiety disorder with
                sensitivity 89%, specificity 82% (Spitzer et al. 2006, PubMed
                ID 16717171);
                panic disorder with sensitivity 74%, specificity 81% (Kroenke
                et al. 2010, PMID 20633738);
                social anxiety with sensitivity 72%, specificity 80% (Kroenke
                et al. 2010);
                post-traumatic stress disorder with sensitivity 66%,
                specificity 81% (Kroenke et al. 2010).
                The majority of evidence contributing to these figures comes
                from primary care screening studies.
            </div>
        """
        return h
